zebra-state/src/config.rs

```diff
@@ -261,8 +280,70 @@ fn parse_dir_name(entry: &DirEntry) -> Option<String> {
 /// Parse the state version number from `dir_name`.
 ///
 /// Returns `None` if parsing fails, or the directory name is not in the expected format.
-fn parse_version_number(dir_name: &str) -> Option<u32> {
+fn parse_version_number(dir_name: &str) -> Option<u64> {
     dir_name
         .strip_prefix('v')
         .and_then(|version| version.parse().ok())
 }
+
+/// Returns the full semantic version of the currently running database format code.
+///
+/// This is the version implemented by the Zebra code that's currently running,
+/// the minor and patch versions on disk can be different.
+pub fn database_format_version_in_code() -> Version {
+    Version::new(
+        DATABASE_FORMAT_VERSION,
+        DATABASE_FORMAT_MINOR_VERSION,
+        DATABASE_FORMAT_PATCH_VERSION,
+    )
+}
+
+/// Returns the full semantic version of the on-disk database.
+///
+/// This is the format of the data on disk, the minor and patch versions
+/// implemented by the running Zebra code can be different.
+pub fn database_format_version_on_disk(
+    config: &Config,
+    network: Network,
+) -> Result<Version, BoxError> {
+    let version_path = config.version_file_path(network);
+
+    let version = match fs::read_to_string(version_path) {
+        Ok(version) => version,
+        Err(e) if e.kind() == ErrorKind::NotFound => {
+            // If the version file doesn't exist, assume both minor and patch are zero.
+            // (The major versions must be the same: we create a new database if they are not.)
+            return Ok(Version::new(DATABASE_FORMAT_VERSION, 0, 0));
+        }
+        Err(e) => Err(e)?,
+    };
+
+    let (minor, patch) = version
+        .split_once('.')
+        .ok_or("invalid database format version file")?;
+
+    Ok(Version::new(
+        DATABASE_FORMAT_VERSION,
+        minor.parse()?,
+        patch.parse()?,
+    ))
+}
+
+/// Writes the currently running semantic database version to the on-disk database.
+/// This should only be called after all running format upgrades are complete.
+pub fn write_database_format_version_to_disk(
+    config: &Config,
+    network: Network,
+) -> Result<(), BoxError> {
+    let version_path = config.version_file_path(network);
+
+    // The major version is already in the directory path.
+    let version = format!(
+        "{}.{}",
+        DATABASE_FORMAT_MINOR_VERSION, DATABASE_FORMAT_PATCH_VERSION
+    );
+
+    fs::write(version_path, version.as_bytes())?;
```

Comment:
Comment needed: we don't need to worry about concurrency here, because this method is only called after RocksDB exclusively locks the state directory.